goal_src/goal-lib.gc

```diff
@@ -376,19 +376,89 @@
   )
 )
 
-(defmacro dotimes (var &key (label #f) &rest body)
-  "Loop like for (int i = 0; i < end; i++)
-   var is a list made up of a variable to bind the amount to (second item), and the remaining forms are evaluated after the loop is finished."
-  (let ((continue-label (string->symbol (if label (string-append (symbol->string label) "-continue") "continue"))))
-    `(let (( ,(first var) 0))
-      (while (< ,(first var) ,(second var)) :label ,label
-              (block ,continue-label
-                ,@(if (null? body) (list `(return-from ,continue-label #f)) body)
-              )
-              (1+! ,(first var))
-              )
-      ,@(cddr var)
+(defmacro dotimes (var &key (label #f) &key (unroll 1) &rest body)
+  "Loop like for (int i = 0; i < end, i++)
+   var is a list made up of a variable to bind the amount to (second item), and the remaining forms are evaluated after the loop is finished
+   Supports optional loop unrolling with :unroll factor
+   Supports (_ x) syntax to iterate without binding and incrementing a variable when unrolled, and unrolls by default
+   Warning: Does not validate the :unroll factor ratio and doesn't support break/continue when unrolling"
+  (let ((index-var (first var))
+        (limit (second var))
+        (continue-label (string->symbol (if label (string-append (symbol->string label) "-continue") "continue")))
+        (unrolled-body '())
+       )
+
+    (when (and (integer? limit)
+               (> unroll limit)
+           )
+      (error ":unroll factor must be <= the loop limit.")
+    )
+    (when (< unroll 1)
+      (error ":unroll factor must be >= 1.")
+    )
```

Comment:
For whatever reason, in the original code they left `dotimes` in that iterate `0` times.  This highlights an edge-case with the assumption that every loop will execute atleast once.

See `cloth.gc` in jak3